dff/context_storages/context_schema.py

```diff
@@ -0,0 +1,286 @@
+"""
+Context Schema
+--------------
+The `ContextSchema` module provides class for managing context storage rules.
+The :py:class:`~.Context` will be stored in two instances, `CONTEXT` and `LOGS`,
+that can be either files, databases or namespaces. The context itself alongsode with
+several latest requests, responses and labels are stored in `CONTEXT` table,
+while the older ones are kept in `LOGS` table and not accessed too often.
+"""
+
+import time
+from asyncio import gather
+from uuid import uuid4
+from enum import Enum
+from pydantic import BaseModel, Field
+from typing import Any, Coroutine, List, Dict, Optional, Callable, Tuple, Union, Awaitable
+from typing_extensions import Literal
+
+from dff.script import Context
+
+ALL_ITEMS = "__all__"
+"""
+The default value for all `DictSchemaField`:
+it means that all keys of the dictionary or list will be read or written.
+Can be used as a value of `subscript` parameter for `DictSchemaField` and `ListSchemaField`.
+"""
+
+_ReadPackedContextFunction = Callable[[str], Awaitable[Tuple[Dict, Optional[str]]]]
+"""
+Type alias of asynchronous function that should be called in order to retrieve context
+data from `CONTEXT` table. Matches type of :py:func:`DBContextStorage._read_pac_ctx` method.
+"""
+
+_ReadLogContextFunction = Callable[[Optional[int], str, str], Awaitable[Dict]]
+"""
+Type alias of asynchronous function that should be called in order to retrieve context
+data from `LOGS` table. Matches type of :py:func:`DBContextStorage._read_log_ctx` method.
+"""
+
+_WritePackedContextFunction = Callable[[Dict, int, int, str, str], Awaitable]
+"""
+Type alias of asynchronous function that should be called in order to write context
+data to `CONTEXT` table. Matches type of :py:func:`DBContextStorage._write_pac_ctx` method.
+"""
+
+_WriteLogContextFunction = Callable[[List[Tuple[str, int, Any]], int, str], Coroutine]
+"""
+Type alias of asynchronous function that should be called in order to write context
+data to `LOGS` table. Matches type of :py:func:`DBContextStorage._write_log_ctx` method.
+"""
+
+
+class SchemaField(BaseModel, validate_assignment=True):
+    """
+    Schema for :py:class:`~.Context` fields that are dictionaries with numeric keys fields.
+    Used for controlling read and write policy of the particular field.
+    """
+
+    name: str = Field(default_factory=str, frozen=True)
+    """
+    `name` is the name of backing :py:class:`~.Context` field.
+    It can not (and should not) be changed in runtime.
+    """
+
+    subscript: Union[Literal["__all__"], int] = 3
+    """
+    `subscript` is used for limiting keys for reading and writing.
+    It can be a string `__all__` meaning all existing keys or number,
+    positive for first **N** keys and negative for last **N** keys.
+    Keys should be sorted as numbers.
+    Default: 3.
+    """
+
+
+class ExtraFields(str, Enum):
```

Comment:
This class helps to keep track of `Context` fields.
Yes, we could've just use field names instead, but that way we would've no way for checking it for e.g. typos.
Using this approach, since all field names are enum elements, we can be sure we make no mistakes while referencing certain fields.